src/backend/cdb/cdbmutate.c

```diff
@@ -1113,6 +1116,375 @@ request_explicit_motion(Plan *plan, Index resultRelationsIdx, List *rtable)
+			{
+				TargetEntry	*tle;
+
+				Assert(IsA(lfirst(lct), TargetEntry));
+
+				tle = (TargetEntry *) lfirst(lct);
+				if (IsA(tle->expr, Var) &&
+					(((Var *) (tle->expr))->varattno == ((Var *) lfirst(lcv))->varattno))
+					break;
+			}
+
+			if (!lct)
+			{
+				newTle = makeTargetEntry((Expr *) lfirst(lcv), targetResno,
+										 "" /* resname */, false /* resjunk */);
+				plan->targetlist = lappend(plan->targetlist, newTle);
+				++targetResno;
+			}
+		}
+		pcontext->found = true;
+
+		return;
+	}
+
+	if (plan->lefttree)
+	{
+		append_absent_targetlist_mutator(plan->lefttree, pcontext);
+
+		if (!pcontext->found && plan->righttree)
+			append_absent_targetlist_mutator(plan->righttree, pcontext);
+
+		if (pcontext->found)
+		{
+			ListCell	*lcv;
+			int			targetResno;
+
+			targetResno = list_length(plan->targetlist) + 1;
+
+			foreach(lcv, pcontext->absent_vars)
+			{
+				Var			*var;
+				ListCell	*lct;
+				TargetEntry *tle;
+
+				Assert(IsA(lfirst(lcv), Var));
+
+				var = (Var *) lfirst(lcv);
+
+				foreach(lct, plan->targetlist)
+				{
+					tle = (TargetEntry *) lfirst(lct);
+
+					if (IsA(tle->expr, Var) &&
+						((Var *) tle->expr)->varno == pcontext->result_relation_idx &&
+						((Var *) tle->expr)->varattno == var->varattno)
+						break;
+				}
+
+				if (!lct)
+				{
+
+					tle = makeTargetEntry(lfirst(lcv), targetResno++, "" /* resname */, false /* junk */);
+					plan->targetlist = lappend(plan->targetlist, tle);
+				}
+			}
+
+			return;
+		}
+	}
+
+	return;
+}
+
+
+/*
+ * Add absent vars to targetlist recursively. We just go into plan node, not subplan or
+ * other kind of nodes.
+ */
+static void
+append_absent_targetlist(struct PlannerInfo *root, Plan *plan,
+						 List *varsAbsent, Index resultRelationsIdx)
+{
+	append_absent_targetlist_context context;
+
+	planner_init_plan_tree_base(&context.base, root);
+
+	context.absent_vars = varsAbsent;
+	context.result_relation_idx = resultRelationsIdx;
+	context.found = false;
+
+	return append_absent_targetlist_mutator(plan, &context);
+}
+
+static void
+failIfUpdateTriggers(Relation relation)
+{
+	bool	found = false;
+
+	if (relation->rd_rel->relhastriggers && NULL == relation->trigdesc)
+		RelationBuildTriggers(relation);
+
+	if (!relation->trigdesc)
+		return;
+
+	if (relation->rd_rel->relhastriggers)
+	{
+		for (int i = 0; i < relation->trigdesc->numtriggers && !found; i++)
+		{
+			Trigger trigger = relation->trigdesc->triggers[i];
+			found = trigger_enabled(trigger.tgoid) &&
+					(get_trigger_type(trigger.tgoid) & TRIGGER_TYPE_UPDATE) == TRIGGER_TYPE_UPDATE;
+			if (found)
+				break;
+		}
+	}
+
+	if (found || child_triggers(relation->rd_id, TRIGGER_TYPE_UPDATE))
+	{
+		ereport(ERROR, (errcode(ERRCODE_GP_FEATURE_NOT_YET),
+						errmsg("Cannot parallelize an UPDATE statement that updates the distribution columns")));
+		relation_close(relation, NoLock);
+	}
+}
+
+/*
+ * In legacy planner, we add a SplitUpdate node at top so that updating on distribution
```

Comment:
This could use a more verbose comment. Some important things to note/explain, either here in the function comment, or in more detailed comments inside the function:

* The table OID, ctid and action are added into the targetlist.
* What are all the args?
* 'mt' is expected to be a ModifyTable. (Perhaps change the datatype to ModifyTable to make that more clear). It is modified with the column indexes of the added ctid/oid/action junk attributes. 
*